language/e2e-testsuite/src/tests/script_functions.rs

```diff
@@ -0,0 +1,140 @@
+                return;
+            }
+
+            public f_public(s: &signer) {
+                return;
+            }
+
+            public(script) f_script(s: &signer) {
+                return;
+            }
+        }
+        ",
+    );
+    let compiled_module = compile_module_with_address(account.address(), "file_name", &program).1;
+
+    let txn = account
+        .transaction()
+        .module(compiled_module)
+        .sequence_number(seq_num)
+        .sign();
+
+    let output = executor.execute_transaction(txn);
+    // module publishing should always succeed
+    assert!(transaction_status_eq(
+        &output.status(),
+        &TransactionStatus::Keep(KeptVMStatus::Executed),
+    ));
+    executor.apply_write_set(output.write_set());
+
+    seq_num + 1
+}
+
+#[test]
+fn script_fn_payload_invoke_private_fn() {
+    let mut executor = FakeExecutor::from_genesis_with_options(VMPublishingOption::open());
+    executor.set_golden_file(current_function_name!());
+
+    let sequence_number = 2;
+    let account = executor.create_raw_account_data(1_000_000, sequence_number);
+    executor.add_account_data(&account);
+
+    let sequence_number = prepare_module(&mut executor, account.account(), sequence_number);
+
+    let txn = account
+        .account()
+        .transaction()
+        .script_function(ScriptFunction::new(
+            ModuleId::new(*account.address(), Identifier::new("M").unwrap()),
+            Identifier::new("f_private").unwrap(),
+            vec![],
+            vec![],
+        ))
+        .sequence_number(sequence_number)
+        .sign();
+
+    let output = executor.execute_transaction(txn);
+    assert!(transaction_status_eq(
+        &output.status(),
+        &TransactionStatus::Keep(KeptVMStatus::MiscellaneousError),
+    ));
+}
+
+#[test]
+fn script_fn_payload_invoke_public_fn() {
+    let mut executor = FakeExecutor::from_genesis_with_options(VMPublishingOption::open());
+    executor.set_golden_file(current_function_name!());
+
+    let sequence_number = 2;
+    let account = executor.create_raw_account_data(1_000_000, sequence_number);
+    executor.add_account_data(&account);
+
+    let sequence_number = prepare_module(&mut executor, account.account(), sequence_number);
+
+    let txn = account
+        .account()
+        .transaction()
+        .script_function(ScriptFunction::new(
+            ModuleId::new(*account.address(), Identifier::new("M").unwrap()),
+            Identifier::new("f_public").unwrap(),
+            vec![],
+            vec![],
+        ))
+        .sequence_number(sequence_number)
+        .sign();
+
+    let output = executor.execute_transaction(txn);
+    assert!(transaction_status_eq(
+        &output.status(),
+        &TransactionStatus::Keep(KeptVMStatus::MiscellaneousError),
+    ));
+}
+
+#[test]
+fn script_fn_payload_invoke_script_fn() {
+    let mut executor = FakeExecutor::from_genesis_with_options(VMPublishingOption::open());
```

Comment:
QQ: What should the DiemVersion be in the initial genesis? I suppose it should be initialized in the `DiemVersion` module so it should be 0 at the beginning? Why is it already activated here?

Reply:
No, it is not activated. See the CI log, in particular, the following tests failed (and this is expected).

```
    tests::script_functions::script_fn_payload_invoke_private_fn
    tests::script_functions::script_fn_payload_invoke_public_fn
    tests::script_functions::script_fn_payload_invoke_script_fn
```

I'll update the e2e tests once the gating logic makes sense to you.

Reply:
The gating logic makes sense! Let's proceed with unit tests first!

Reply:
Coming up soon

Reply:
Done, the e2e test cases are updated.

- The feature is gated prior to `set_diem_version()`, and hence, the transaction finishes with `TransactionStatus::Discard(DiscardedVMStatus::FEATURE_UNDER_GATING)`.
- After the version update, the transaction will proceed and results in either 
  - `TransactionStatus::Keep(KeptVMStatus::MiscellaneousError)` if not calling a `public(script)` function, or
  - `TransactionStatus::Keep(KeptVMStatus::Executed)` if everything is OK.